### DISCUSSION TITLE
use IfcOpenShell to draw ifc in OpenGL

### DISCUSSION THREAD

**Discussion by Ttofuuu** [IfcOpenShell/IfcOpenShell#1933] [Q&A] [answered]
Currently I've builded IfcOpenShell ver 0.7.0, and tried ifcConvert.exe to trans an ifc file into obj file. The conversion works great but I have trouble in the next step: my goal is to read an ifc file, and draw it through OpenGL. So my problem comes up: how can I get the geometry data from ifc, and trans them to OpenGL vertex buffer object directly, instead of transing them into obj then read obj into vertex buffer object? I wonder if there is a obvious data stream we can cite to boost my target.
I'll appreciate it if somebody know that or show me codes about this. Too fresh to be a coder that I nearly have no ability to handle this🤣🤣

**Comment by aothms** [comment, marked as answer]
IfcConvert is just one possible use of the "geometry iteration" mechanism we have in IfcOpenShell. Using `IfcGeom::Iterator` which you can iterate over to give you a bunch of `IfcGeom::TriangulationElement`. The data model is something you can almost directly feed to OpenGL.

If you want I think it'd make a nice tutorial series for academy.ifcopenshell.org/ so I can help code review and provide feedback if we get a tutorial out of it.

**Reply by Ttofuuu** [reply]
Thanks aothms for replying, I've read some codes of `IfcGeom`, and have some questions still.
1. It seems that the `IfcGeom` would create a lib, depended on `IfcParse`, `IfcGeom_ifc2x3`, `IfcGeom_ifc4x1` etc. So can I just include the `IfcGeom.lib` to use the class like `IfcGeom::Iterator` in a new VS solution? If I can't, how can I edit the source code to create a lib to use?
2. To initial an `IfcGeom::Iterator`, `IfcGeom::IteratorSettings& settings`, `IfcParse::IfcFile* file` should be offered as parameters, I guess the `file` is the Ifc file to be transed into geometry data, but how can I create an `IfcGeom::IteratorSettings`?
3. It seems like I can create an `IfcGeom::Iterator`, and use the inner function `IfcGeom::Iterator::get()` to create `IfcGeom::TriangulationElement` , and use `IfcGeom::TriangulationElement::geometry()` to get my goal, right?
Really glad to see your reply before, it really helps a lot. Looking forward to your reply again!

**Reply by aothms** [reply]
Your assumptions are correct. For iterator settings you just instantiate it and then toggle features on an off using `set()`, eg. `settings.set(IfcGeom::IteratorSettings::WELD_VERTICES, false)`

**Reply by Ttofuuu** [reply]
Thanks for showing me how to create an `IfcGeom::IteratorSettings`!

> Your assumptions are correct.

What about question one? Am I able to use the inner function and class after including `IfcGeom.lib`?

**Reply by aothms** [reply]
The dependency information is not carried over, so you do need to add the other libraries to your linker input.

**Reply by Ttofuuu** [reply]
Thank you aothms, I'll test it soon.

**Reply by Ttofuuu** [reply]
Sorry for it taking so long to reply! For convenience, I directly take the sample and modify it. However, I found it whichever .ifc file I offered, the `IfcGeom::Representation::Representation`'s verts remain the length of 72. Am I understanding something wrong? Is anything should be done with the `iterator`, and how?
Here is my code. It seems the .cpp file could not be uploaded in reply, so I trans it into a .txt file.
[temp.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/7834561/temp.txt)

**Reply by aothms** [reply]
> verts remain the length of 72

Note that that's the amount of vertex components (x y and z), so 24 vertices. That's 8 vertices along 3 different faces. This is the most common BIM element, e.g wall, rectangular slab will all have that amount of verts. So it's a coincidence, but not unlikely.

In you code you don't actually traverse the iterator. If you create a loop using next() you'll get more different data.

**Reply by Ttofuuu** [reply]
Thanks for the explanation, I understand and the code does run successfully!

**Reply by Ttofuuu** [reply]
Hey sir, currently I found a problem. When I'm trying to generate a building into triangles, something wrong happened.
![image](https://user-images.githubusercontent.com/49158715/150677203-2339141e-7ac1-4c90-bb11-f80f853259e7.png)
This is the building ifc file, opened in xplorer. When I want to generate it into triangles to draw through opengl, I found that the component of IfcBeam could be draw correctly, but when it comes to IfcWall, IfcSpace or other components, they failed to generate the right position, looks like the triangles we generated all come to the lawest floor:
![PJ RF9UG7G@NQGE@_ZJNDWD](https://user-images.githubusercontent.com/49158715/150677362-1942d11b-b437-4412-a30b-5e5572ec46ee.JPG)
But when we use IfcConvert to trans it into obj, it gives the right answer. So I began to suspect that I was using it wrong. Here comes my code in .txt format, looking forward to your reply. 
An
[tempv2.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/7920416/tempv2.txt)
d, happy new year!

**Reply by aothms** [reply]
If you use the code like that you get elements in their local coordinate system (meaning that for ex the same windows can reuse the same geometry definition). Quickest way to get the correct output is to set

`iteratorSettings.set(IfcGeom::IteratorSettings::USE_WORLD_COORDS, true);`

That way element geometry is in absolute world coordinates, the ability for reuse is eliminated, so it's not the most efficient memory wise.

**Reply by Ttofuuu** [reply]
So I need to add `iteratorSettings.set(IfcGeom::IteratorSettings::USE_WORLD_COORDS, true);` in my code, and keep `iteratorSettings.set(IfcGeom::IteratorSettings::WELD_VERTICES, false);` this line as well, right?

**Reply by aothms** [reply]
yep :)

**Reply by Ttofuuu** [reply]
haha thank you sir!

**Reply by Ttofuuu** [reply]
Hey sir, I wonder while using  `IfcGeom::Iterator` to generate components, can I find out which floor the component on?

**Reply by aothms** [reply]
Using parent_id and get_object you can iterate until you reach an IfcBuildingStorey

https://ifcopenshell.github.io/docs/rst_files/class_ifc_geom_1_1_element.html#_CPPv4NK7IfcGeom7Element9parent_idEv

https://ifcopenshell.github.io/docs/rst_files/class_ifc_geom_1_1_iterator.html#_CPPv4N7IfcGeom8Iterator10get_objectEi

**Reply by Ttofuuu** [reply]
Great, but after finding out the IfcBuildingStorey element, from which function of IfcGeom::Element could I know which floor the IfcBuildingStorey element represent? Seems like the `product() ` func has the ability to do that, but I haven't test it, because  I don't know how to use it😢

**Reply by Ttofuuu** [reply]
by the way, how can I get the properties data of one component? It looks like they are same question.

**Reply by Ttofuuu** [reply]
> Great, but after finding out the IfcBuildingStorey element, from which function of IfcGeom::Element could I know which floor the IfcBuildingStorey element represent? Seems like the `product() ` func has the ability to do that, but I haven't test it, because I don't know how to use it😢

I find the answer in `Element->name()` hahaha, thanx sir!!


> by the way, how can I get the properties data of one component? It looks like they are same question.

but still dont know where the properties are

**Reply by aothms** [reply]
Properties follow this structure http://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/concepts/Object_Definition/Property_Sets/Property_Sets_for_Objects/content.html

Starting from the left: IsDefinedBy -> RelatingPropertyDefinition -> HasProperties -> {Name, NominalValue}.

product() is a latebound entity definition not tied to a specific version of the schema.

So you need to do something like:

~~~C++
auto rels = element->product()->get_inverse("IsDefinedBy");
for (auto& rel : *rels) {
  IfcUtil::IfcBaseClass* inst = rel->get("RelatingPropertyDefinition");
}
~~~

If you know the schema, it's a bit easier to begin with, like:

~~~C++
auto rels = element->product()->as<Ifc2x3::IfcElement>()->IsDefinedBy();
for (auto& rel : *rels) {
  auto pset = rel->RelatingPropertyDefinition();
}
~~~

http://blog.ifcopenshell.org/2019/12/v060.html

**Reply by Ttofuuu** [reply]
Wooh, really appreciate it for your detailed answers! I'll check the blog and codes soon.
There's another question: while generating triangles, I find that some triangles, which never appear on the obj generated by IfcConvert, whould be generated to be visable, and I'm surely that I've filtered some triangles out by generating specified components. I wonder where do these tirangles come from,  and how IfcConvert fliters them out.
Here comes the example:
this is the model generated by IfcConvert:
![2b3b8d4b4005b82b8dcb86f30cc7627](https://user-images.githubusercontent.com/49158715/156146763-93c19c89-243e-4248-86a4-e4b65023ad78.png)

this is the triangles generated by my code:
![827654c79155c7b8fc95e0dcc582572](https://user-images.githubusercontent.com/49158715/156143980-702782c9-65e8-4853-9151-41a362243053.png)

Looks like I should change my IteratorSettings to be same with the IteratorSettings in the IfcConvert code, is this right?

here comes my code:
[TempCode.txt](https://github.com/IfcOpenShell/IfcOpenShell/files/8160589/TempCode.txt)

Looking forward to your replying. Your answer is always helpful.

**Reply by aothms** [reply]
Hard to say without more knowledge of the specific content. Is it just a different element type? (IfcOpeningElement, IfcSpace) these are filtered out by default in IfcConvert.

**Reply by Ttofuuu** [reply]
Seems not. Looks like the IfcBeams and IfcColumns would duplicate theirself, then rotate and transform to the wrong place. I'd like to share my temp Ifc file, it works good through IfcConvert, but really bad on my code.
[temp.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/8167347/temp.zip)

**Reply by Ttofuuu** [reply]
A minute, recently I regenerated the whole ifcOpenShell, and now I find the IfcConvert has the same problem😲 so I need to redownload the whole thing right?

**Reply by aothms** [reply]
Is it possible that these are the representations with IfcBoundingBox? We recently had another user complain about that as well.

https://github.com/IfcOpenShell/IfcOpenShell/issues/2044

And I see they are in your model such as #111= IFCBOUNDINGBOX(#110,600.,750.,7799.998046875);                                                       

They were added fairly recently.

https://github.com/IfcOpenShell/IfcOpenShell/pull/1743

**Reply by Ttofuuu** [reply]
It seems not. I add this line on line1431 in IfcConvert.cpp file:
`entities.insert("IfcBoundingBox");`
to filter the IfcBoundingBox away, but the IfcConvert still generate some ugly components out, like tha:
which it should be:
![1646318442(1)](https://user-images.githubusercontent.com/49158715/156586933-491937ef-fe08-4a57-bbfc-0b0c9f663e89.png)
which ifcConvert(locally compiled) generated:
![1646318493(1)](https://user-images.githubusercontent.com/49158715/156587071-958f5a5d-7340-4355-ab09-454912b5977c.png)
but I've just downloaded the latest ifcConvert.exe, and it works really good:
![1646318680(1)](https://user-images.githubusercontent.com/49158715/156587666-ac138066-65fd-4599-b376-5db163c3c1db.png)
so I really wonder how this mistake happens.

**Reply by aothms** [reply]
>  entities.insert("IfcBoundingBox");

it wouldn't work like that. the filter works for the building element subtypes, but IfcBoundingBox is the representation item type (the geometry) referred to by the building element.

**Reply by Ttofuuu** [reply]
So I should judge if the `IfcGeom::TriangulationElement`'s geometry() is the IfcBoundingBox, then could I filter it out if it is, right?

**Reply by Ttofuuu** [reply]
> Using parent_id and get_object you can iterate until you reach an IfcBuildingStorey
> 
> https://ifcopenshell.github.io/docs/rst_files/class_ifc_geom_1_1_element.html#_CPPv4NK7IfcGeom7Element9parent_idEv
> 
> https://ifcopenshell.github.io/docs/rst_files/class_ifc_geom_1_1_iterator.html#_CPPv4N7IfcGeom8Iterator10get_objectEi

There's another question: when I use 
``` cpp
while (cur_element->type()!= "IfcBuildingStorey") { // find current parent
   cur_element = iterator.get_object(cur_element->parent_id());
}
```
to get the floor, I find it will crush in this ifc file when the `cur_lement->parent_id()` is a floor's id,e.g. `cur_lement->parent_id()`=26. But things go well when it turns to other ifc files. Why would this happen?
here's the crushing ifc file:
[crushing floor.zip](https://github.com/IfcOpenShell/IfcOpenShell/files/8180302/crushing.floor.zip)

**Reply by Ttofuuu** [reply]
I have too many questions to ask! If I get one IfcBuildingStorey element now, how can I get the elevation? And when it comes to an IfcExtrudedSolid, how can I get the geom information of it?

**Reply by aothms** [reply]
It's not a given that you're going to find a building storey. Did you try this:

~~~C++
while (cur_element->type()!= "IfcBuildingStorey" && cur_element->parent_id()) { 
   ...
~~~

> how can I get the elevation

It has an elevation attribute. *if and only if it is a buildingstorey*

~~~C++
auto attr = cur_element->product->get("Elevation");
if (!attr->isNull()) {
  double elev = *attr;
}
~~~

> how can I get the geom information of it

What information do you need? You already have the triangulated data.

**Reply by Ttofuuu** [reply]
> Did you try this

Currently I'm using code like this and it works perfectly in IfcOpenShell v0.6.0, thanks again for your team's code work.


>It has an elevation attribute. if and only if it is a buildingstorey

Great, I'll test it soon.

>What information do you need? You already have the triangulated data.

You are right, I already have the triangulated data. In fact I wonder if I could get the length/width/height from an element, which is a cuboid, described by IfcExtrudedAreaSolid and has no property of length/width/height. Maybe I can find my way out from the boundingbox of this element.

Anyway, thank you again for your prompt reply!

**Reply by aothms** [reply]
> which is a cuboid, described by IfcExtrudedAreaSolid and has no property of length/width/height

Yes this is always the choice you have to make. Slabs are generally extrusions, so you might assume it's an IfcExtrudedAreaSolid, but it doesn't have to be. So reverse engineering those parameters from triangulation (or BRepElement using OpenCASCADE would be preferable but another API to learn) is a general solution that always works, but can be a bit painful to implement.

**Reply by Ttofuuu** [reply]
>  or BRepElement using OpenCASCADE would be preferable but another API to learn

Do you mean [this](https://ifcopenshell.github.io/docs/rst_files/class_ifc_geom_1_1_b_rep_element.html?highlight=brepelement#_CPPv4NK7IfcGeom11BRepElement8geometryEv) sir?

**Reply by aothms** [reply]
Yes, and then by calling as_compound() you get a single OpenCASCADE element for a product, that you can use for all possible analysis https://ifcopenshell.github.io/docs/rst_files/class_ifc_geom_1_1_representation_1_1_b_rep.html#_CPPv4NK7IfcGeom14Representation4BRep11as_compoundEb

**Reply by Ttofuuu** [reply]
> that you can use for all possible analysis

I'll read about `TopoDS_Compound` and find what it is soon. But before that, let me show you some paragraph in an ifc file:
```
#14961= IFCCOLUMN('a1a813ce-1a41-46c0-b816-ca39dcb8c376',#5,'\X2\77E95F6267F1\X0\',$,$,#14960,#14957,$); 

#14963= IFCPROPERTYSINGLEVALUE('\X2\5BBD5EA60042\X0\',$,IFCLABEL('\X2\003800300030\X0\'),$);   
#14964= IFCPROPERTYSINGLEVALUE('\X2\9AD85EA60048\X0\',$,IFCLABEL('\X2\003400300030\X0\'),$);   
#14965=IFCPROPERTYSET('4fc8e74c-2a29-4192-8f48-8657861db31b',#5,'\X2\53C26570\X0\',$,(#14963,#14964));
#14966= IFCRELDEFINESBYPROPERTIES('ec09a468-3ba2-4495-8fca-da16a807ae1e',#5,$,$,(#14961),#14965);  
```
I'm using code like this:
```
auto rels = this_ifc_column_element->product()->get_inverse("IsDefinedBy");
for (auto& rel : *rels) {
  IfcUtil::IfcBaseClass* inst = rel->get("RelatingPropertyDefinition");
}
```
Then what should I do next to get the `\X2\5BBD5EA60042\X0\`'s data in #14963 line? 
Learnt a lot from your replies already, really thanks.

**Reply by aothms** [reply]
You still need to do `HasProperties` (returns `aggregate_of_instance::ptr`) loop over these to get `NominalValue` (returns `IfcUtil::IfcBaseClass*`) and then just call `attr = nv.data.getArgument(0)`. Check `attr->type() == Argument_STRING`. `std::string attr_value = *attr;
